### PR TITLE
Error Handling and further cleanup

### PR DESCRIPTION
**Pull request by Nereuxofficial**
While the examples do successfully compile but do not really work yet, there are some further Cleanups along with Changes like deriving the Debug trait to many structs which makes it possible to use it on types like the Snake in the example.

I also made an error.rs for the train module which makes error handling possible and is shorter than using panic twice.

### CODE REVIEW

File: examples/snake-cli/coordinate.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                }
+            }
+        }
+    }
+    pub fn new_transform(
+        coordinate: Coordinate,
+        direction: Direction,
+        step: usize,
+    ) -> (Self, bool) {
+        let mut clone = coordinate;
```
Comment (sakex) [inline examples/snake-cli/coordinate.rs:51]: You could have the `mut` at line 47

File: examples/snake-cli/game.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    scores: Vec<f64>,
+}
+
+impl Game {
+    pub fn get_scores(&self) -> Vec<f64> {
+        self.scores.clone()
+    }
+
+    pub fn new(neural_networks: Vec<NeuralNetwork<f64>>) -> Game {
+        let mut snakes = vec![];
```
Comment (sakex) [inline examples/snake-cli/game.rs:22]: `let snakes = neural_networks.iter().map(Snake::new).collect()`

File: examples/snake-cli/game.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            self.update();
+        }
+    }
+
+    /// Make the snakes make their decision
+    pub fn make_decision(&mut self) {
+        let mut inputs: [f64; 5] = [0., 0., 0., 0., 0.];
+        let cloned_apple = self.apple.clone();
+        // Let each snake make a decision
+        self.snakes.iter_mut().for_each(|s| {
```
Comment (sakex) [inline examples/snake-cli/game.rs:47]: You can consider multithreading that in the future with `rayon`

File: examples/snake-cli/game.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            inputs[0] = distance_to_apple_x(s, cloned_apple);
+            inputs[1] = distance_to_apple_y(s, cloned_apple);
+            inputs[2] = distance_to_wall_x(s);
+            inputs[3] = distance_to_wall_y(s);
+            s.make_decision(&inputs)
+        });
+    }
+
+    /// Updates the game. Should be called every tick
+    pub fn update(&mut self) {
```
Comment (sakex) [inline examples/snake-cli/game.rs:58]: The problem is that you are moving the apple for all snakes at once when you would want to have one apple per snake

File: examples/snake-cli/game.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        // Let each snake make a decision
+        self.make_decision();
+    }
+
+    fn game_over(&self) -> bool {
+        self.snakes.is_empty()
+    }
+
+    fn remove_if_dead(&mut self) {
+        for idx in 0..self.snakes.len() {
```
Comment (sakex) [inline examples/snake-cli/game.rs:79]: Could zip `snakes` and `scores`

File: examples/snake-cli/game.rs (hunk 4 above)
Comment (sakex) [reply examples/snake-cli/game.rs:58]: (You want to replace only if the snake ate the apple for each snake)

File: examples/snake-cli/snake.rs [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+pub struct Snake {
+    blocks: LinkedList<Block>,
+    moving_direction: Direction,
+    dir_changed: bool,
+    net: Option<NeuralNetwork<f64>>,
+}
+
+impl Snake {
+    pub fn new(net: NeuralNetwork<f64>) -> Self {
+        let mut blocks = LinkedList::new();
```
Comment (sakex) [inline examples/snake-cli/snake.rs:23]: I usually try to avoid `LinkedList` since they are very inefficient for the cache, but it is very elegant here

File: src/utils.rs [hunk 7]
```diff
@@ -4,5 +4,11 @@ use numeric_literals::replace_numeric_literals;
 #[inline(always)]
 #[replace_numeric_literals(T::from(literal).unwrap())]
 pub fn floats_almost_equal<T: Float>(f1: T, f2: T) -> bool {
-    f1 + 1e-7 > f2 && f1 - 1e-7 < f2
```
Comment (sakex) [inline src/utils.rs:7]: Why the change here? It was valid before

File: src/tests.rs [hunk 8]
```diff
@@ -98,23 +97,22 @@ impl Game<f64> for TestGame {
...
                 println!("{:?}, {:?}", top.layers_sizes, top2.layers_sizes);
                 println!("{}", as_str);
-                println!("========================================");
+                section!();
                 println!("{}", top2.to_string());
                 panic!("Network != Network from string");
             }
             self.nets = vec![network, network_from_string];
             let output = self.run_generation();
-            if !(output[0] - 1e-8 < output[1] && output[0] + 1e-8 > output[1]) {
```
Comment (sakex) [inline src/tests.rs:117]: Same comment as in `utils.rs`

File: src/utils.rs (hunk 7 above)
Comment (sakex) [reply src/utils.rs:7]: We are checking if 2 floats are almost equal, up to floating point error

if `f1 - f2 < 1 e-7`, `f1` could be `0` and `f2` `300000`. The original implementation puts the difference between `f1 - f2` in the interval `{-1e-7; 1e-7}`

**Review by sakex — CHANGES_REQUESTED** [review]
Thanks for the huge work! It looks great, I had a few minor comments. I'm really looking forward to seeing the snake run. Are you planning on making a `WASM/desktop` UI?

File: examples/example.rs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        &[[0.0, 0.0], [1.0, 1.0], [1.0, 0.0], [0.0, 1.0]]
+    }
+}
+
+/// Computes the score with given inputs and one output
+fn compute_score(inputs: &[f64], output: f64) -> f64 {
+    // https://en.wikipedia.org/wiki/XOR_gate
+    // Returns 1.0 for a wrong output and 0.0 for a right output. Should be used as a score
+    let inputs: Vec<&f64> = inputs.into_iter().collect_vec();
+    if (*inputs[0] == 0.0 && *inputs[1] == 1.0) || (*inputs[0] == 1.0 && *inputs[1] == 0.0) {
```
Comment (sakex) [inline examples/example.rs:70]: Can be reduced to *inputs[0] ^ *inputs[1] == output

File: examples/example.rs [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+impl Player {
+    pub fn new(net: NeuralNetwork<f64>) -> Player {
+        Player { net }
+    }
+    /// Runs all the inputs and calculates the outputs
+    fn run(&mut self) -> f64 {
+        // Get the inputs
+        let inputs = XOR::get_inputs();
```
[the remark this answers is not in the document]
Comment (Nereuxofficial) [reply examples/example.rs:19]: Something doesn't quite work here... I'm not sure why though

File: examples/example.rs [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        // Calculate a score for every input
+        let outputs: Vec<f64> = inputs.iter().map(|i| self.net.compute(i)[0]).collect();
+        let mut scores: Vec<f64> = vec![];
+        for idx in 0..(inputs.len() - 1) {
+            scores.push(compute_score(&inputs[idx], outputs[idx]));
```
Comment (sakex) [inline examples/example.rs:24]: You need to round the output because it will never be exactly `1` from a `NN` output

Comment (sakex) [reply examples/example.rs:24]: You could also consider having 2 outputs and considering the first one as 0 and the second one as 1, then consider the output as `index(max(output[0], output[1]))`

File: examples/snake-cli/game.rs (hunk 4 above)
Comment (Nereuxofficial) [reply examples/snake-cli/game.rs:58]: You're right but i thought it might be fun to have the snakes race for one Apple instead of having multiple apples

File: examples/snake-cli/coordinate.rs [hunk 12]
```diff
@@ -0,0 +1,60 @@
+use crate::direction::Direction;
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct Coordinate {
+    pub x: usize,
+    pub y: usize,
+}
+impl Coordinate {
+    /// This function returns true if it causes an over-/underflow upon which the snake should die
+    pub fn transform(&mut self, direction: Direction, step: usize) -> bool {
```
[the remark this answers is not in the document]
Comment (Nereuxofficial) [reply examples/snake-cli/coordinate.rs:9]: I should probably let this return a Result<(), OutOfBoundserror> since the transformation only occurs when it is either an under/Overflow.

File: examples/snake-cli/coordinate.rs (hunk 1 above)
Comment (Nereuxofficial) [reply examples/snake-cli/coordinate.rs:51]: I think you're talking about the formatting. Yeah here it is kind of unnecessary how many new lines there are but overall i think cargo fmt does a good job for autoformatting

File: examples/snake-cli/game.rs (hunk 2 above)
Comment (Nereuxofficial) [reply examples/snake-cli/game.rs:22]: Good Cleanup!

File: examples/snake-cli/game.rs (hunk 3 above)
Comment (Nereuxofficial) [reply examples/snake-cli/game.rs:47]: That's probably a good idea although right now i want to get it to work with one thread

File: src/utils.rs (hunk 7 above)
Comment (Nereuxofficial) [reply src/utils.rs:7]: My bad, i thought this did the same and was easier to read

File: examples/example.rs (hunk 9 above)
Comment (Nereuxofficial) [reply examples/example.rs:70]: Yup. Though i should convert them to a boolean first as well as the output which would probably also take care of the Rounding issue

**Comment by Nereuxofficial** [comment]
Thinks for the feedback! As for the desktop/Wasm UI i will first implement it properly then use a UI although i've never used WebAssembly this would probably be a good opportunity to learn it

**Comment by Nereuxofficial** [comment]
Thanks for the rounding tip! Now the XOR example works as you can probably see in the Checks

File: src/topology/gene.rs [hunk 13]
```diff
@@ -239,9 +239,14 @@ where
...
     fn cmp(&self, other: &Self) -> Ordering {
-        if self.output.layer == other.output.layer && self.output.index == other.output.index{
+        if self.output.layer == other.output.layer && self.output.index == other.output.index {
             Ordering::Equal
-        } else if self.output.layer < other.output.layer || (self.output.layer == other.output.layer && self.output.layer < other.output.layer) {
+        } else if self.output.layer < other.output.layer
+            //FIXME: This is impossible
+            || (self.output.layer == other.output.layer && self.output.layer < other.output.layer)
+        // Shouldn't this be:
+        // self.output.layer < other.output.layer && self.output.index < other.output.index
```
Comment (sakex) [inline src/topology/gene.rs]: Correct, nice catch

File: src/topology/gene.rs [hunk 14]
```diff
@@ -239,9 +239,11 @@ where
     T: Float,
 {
     fn cmp(&self, other: &Self) -> Ordering {
-        if self.output.layer == other.output.layer && self.output.index == other.output.index{
+        if self.output.layer == other.output.layer && self.output.index == other.output.index {
             Ordering::Equal
-        } else if self.output.layer < other.output.layer || (self.output.layer == other.output.layer && self.output.layer < other.output.layer) {
+        } else if self.output.layer < other.output.layer
+            || (self.output.layer == other.output.layer && self.output.index < other.output.index)
```
Comment (sakex) [inline src/topology/gene.rs:245]: I corrected it thank you!

**Review by sakex — APPROVED** [review]
LGTM, feel free to merge when you are done!
